### PR TITLE
[Java] Replace binary rewrite with Remote Lambda Cache (SerdeLambda)

### PR DESCRIPTION
**Pull request by mylinyuzhi**
## What do these changes do?
in java-worker, there are two ways to implement the ray.call: asm-rewrite and remote-lambda.
the asm-rewrite is not elegant and remote-lambda is too slow.  And can not get the meta-information of the called function at runtime in the two solutions.  e.g.  @RayRemote(xxx=yyy), at runtime we want to get the xxx=yyy.

the new solution: SerdeLambda( see org.ray.util.LambdaUtils). It has the following advantages:
- support fo get the lambda' meta-information in runtime.
- no need to rewrite the jar. Easy to develop and test.
- one solutuion. the code is simpler
- fast enough(see LambdaUtilsTest): in mac-air 1.6G I5
  - hit the lambda cache, parse the lambda spent 400us+, 80x than remote-lambda
  - miss the lambda cache ,  parse the lambda spent 2600us+, 200x than remote-lambda

### CODE REVIEW

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.WeakHashMap;
+import org.objectweb.asm.Type;
+import org.ray.util.logger.RayLog;
+
+
+public final class MethodId {
+
+  /* use ThreadLocal to avoid lock. </br>
```
Comment (pcmoritz) [inline java/common/src/main/java/org/ray/util/MethodId.java]: unmatched </br>?

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.WeakHashMap;
+import org.objectweb.asm.Type;
+import org.ray.util.logger.RayLog;
+
+
+public final class MethodId {
```
Comment (pcmoritz) [inline java/common/src/main/java/org/ray/util/MethodId.java]: Can we add a small comment on what this class is doing? I.e. it seems to be similar to our FunctionID in Ray, except it also has a class associated with it? Also, let's document the fields methodDesc and encoding, it's not clear to me at the moment what they are.

File: java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java [hunk 3]
```diff
@@ -106,13 +76,47 @@ public synchronized void removeApp(UniqueID driverId) {
...
+      if (tasks == null) {
+        tasks = RayTaskMethods.fromClass(className, classLoader);
+        RayLog.core.info("create RayTaskMethods:" + tasks);
+        taskMethods.put(className, tasks);
+      }
+      RayMethod m = tasks.functions.get(methodId);
+      if (m != null) {
+        return m;
+      }
+      //it is a actor static func
```
Comment (pcmoritz) [inline java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java]: there should be a space after the //

File: java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java [hunk 4]
```diff
@@ -29,74 +29,44 @@ public LocalFunctionManager(RemoteFunctionManager remoteLoader) {
...
-        String logInfo =
-            "DriverId" + driverId + " load remote function " + m.getName() + ", hash = " + k
-                .toString();
-        RayLog.core.debug(logInfo);
-        System.err.println(logInfo);
-        funcs.functions.put(k, v);
-      }
+  Pair<ClassLoader, RayMethod> getMethod(UniqueID driverId, UniqueID actorId,
+      UniqueID methodId, String className) {
+    //assert the driver's resource is load
```
Comment (pcmoritz) [inline java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java]: space after //

File: java/runtime-common/src/main/java/org/ray/core/ArgumentsBuilder.java [hunk 5]
```diff
@@ -83,7 +84,9 @@ private static boolean checkSimpleValue(Object o) {
   @SuppressWarnings({"rawtypes", "unchecked"})
   public static Pair<Object, Object[]> unwrap(TaskSpec task, Method m, ClassLoader classLoader)
       throws TaskExecutionException {
-    FunctionArg[] fargs = task.args;
+    //the last arg is className
```
Comment (pcmoritz) [inline java/runtime-common/src/main/java/org/ray/core/ArgumentsBuilder.java]: space after //

File: java/runtime-native/src/main/java/org/ray/spi/impl/JarLoader.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    Collection<File> jars = FileUtils.listFiles(
+        new File(dir),
+        new RegexFileFilter(".*\\.jar"),
+        DirectoryFileFilter.DIRECTORY
+    );
+    return loadJar(jars, explicitLoad);
+  }
+
+  public static void unloadJars(ClassLoader loader) {
+    // TODO:
```
Comment (pcmoritz) [inline java/runtime-native/src/main/java/org/ray/spi/impl/JarLoader.java]: say what is to be done here (or if nothing, remove the todo)

**Comment by pcmoritz** [comment]
Thanks, this is really great! Some high level questions:

- How do performance numbers you cite compare to the old solution with the binary rewrite? (this solution is clearly superior but it would be good to know the difference)
- Are there any changes in semantics for the user? The changes in https://github.com/ray-project/ray/issues/2172 are orthogonal to this, right?

Also if functions are garbage collected from the WeakDictionary, they will be reloaded later, right? Where are the remote lambdas stored?

File: java/common/src/main/java/org/ray/util/MethodId.java (hunk 2 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: ok, I would add the doc

File: java/common/src/main/java/org/ray/util/MethodId.java (hunk 1 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: ok, I would clean it.

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  /* use ThreadLocal to avoid lock. </br>
+   * A cache from the lambda instances to SerializedLambda.
+   * Note: the lambda instances are dynamically created per call site,
+   * we use WeakHashMap to avoid OOM.
+   */
+  private static final ThreadLocal<WeakHashMap<Class<Serializable>, MethodId>>
+      CACHE = ThreadLocal.withInitial(() -> new WeakHashMap<>());
+
+  /**
+   * format A.B.C.cname
```
[the remark this answers is not in the document]
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: ok, I would clean it.

File: java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java (hunk 3 above)
Comment (mylinyuzhi) [reply java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java]: ok, I would clean it.

File: java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java (hunk 4 above)
Comment (mylinyuzhi) [reply java/runtime-common/src/main/java/org/ray/core/LocalFunctionManager.java]: ok, I would clean it.

File: java/runtime-common/src/main/java/org/ray/core/ArgumentsBuilder.java (hunk 5 above)
Comment (mylinyuzhi) [reply java/runtime-common/src/main/java/org/ray/core/ArgumentsBuilder.java]: ok, I would clean it.

File: java/runtime-native/src/main/java/org/ray/spi/impl/JarLoader.java (hunk 6 above)
Comment (mylinyuzhi) [reply java/runtime-native/src/main/java/org/ray/spi/impl/JarLoader.java]: ok, I would add the doc

File: .gitignore [hunk 8]
```diff
@@ -147,3 +147,9 @@ build
 
 # Vscode
 .vscode/
+
+*.iml
+java/test
+java/*/target
+java/target/
```
Comment (raulchen) [inline .gitignore]: Don't ignore `java/test`, it's a valid module.
`java/test` and `java/*/target` can be combined as `java/**/target`

File: java/common/src/main/java/org/ray/util/LambdaUtils.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * see http://cr.openjdk.java.net/~briangoetz/lambda/lambda-translation.html.
+ */
+public final class LambdaUtils {
+
+  private LambdaUtils() {
+  }
+
+
+  public static SerializedLambda getSerializedLambda(Serializable lambda) {
```
Comment (raulchen) [inline java/common/src/main/java/org/ray/util/LambdaUtils.java]: Let's also use `RayFunc` as the parameter type here.

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+
+public final class MethodId {
+
+  /* use ThreadLocal to avoid lock. </br>
+   * A cache from the lambda instances to SerializedLambda.
+   * Note: the lambda instances are dynamically created per call site,
+   * we use WeakHashMap to avoid OOM.
+   */
+  private static final ThreadLocal<WeakHashMap<Class<Serializable>, MethodId>>
```
Comment (raulchen) [inline java/common/src/main/java/org/ray/util/MethodId.java]: same here, `RayFunc`. 
and a lot of other places in the rest of this PR.

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+  public MethodId(String className, String methodName, String methodDesc, boolean isStatic) {
+    this.className = className;
+    this.methodName = methodName;
+    this.methodDesc = methodDesc;
+    this.isStatic = isStatic;
+    this.encoding = encode(className, methodName, methodDesc, isStatic);
+    this.digest = getSha1Hash0();
+  }
+
```
Comment (raulchen) [inline java/common/src/main/java/org/ray/util/MethodId.java]: nit: remove empty line

File: java/common/src/main/java/org/ray/util/MethodId.java [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    Preconditions.checkArgument(!(serial instanceof SerializedLambda), "arg could not be "
+        + "SerializedLambda");
+    Class<Serializable> clazz = (Class<Serializable>) serial.getClass();
+    WeakHashMap<Class<Serializable>, MethodId> map = CACHE.get();
+    MethodId id = map.get(clazz);
+    if (id == null || forceNew) {
+      final SerializedLambda lambda = LambdaUtils.getSerializedLambda(serial);
+      Preconditions.checkArgument(lambda.getCapturedArgCount() == 0, "could not transfer a lambda "
+          + "which is closure");
+      final boolean isstatic = lambda.getImplMethodKind() == MethodHandleInfo.REF_invokeStatic;
```
Comment (raulchen) [inline java/common/src/main/java/org/ray/util/MethodId.java]: nit: `isStatic`

**Review by raulchen — CHANGES_REQUESTED** [review]

**Comment by raulchen** [comment]
@pcmoritz 
1. The perf of this solution is slightly worse than binary rewrite, but should be totally good enough for prod (see commit message).
2. If everybody is okay with this PR, we'll go on with option 1.
3. Yes, If a function's GC-ed, it will be reloaded. (To be more specific, we're not caching the function, we're caching the metadata of the function).

**Comment by mylinyuzhi** [comment]
## performance
binary rewrite is fastest. the functioId has computed in rewrite. No time overhead for functioId get.
the serdeLambda need to compute the functioInd when miss cache. it need to parse the lambda and sha1 the MethodId. in my mac air, it need 2000+ ns. if hit the cache, it only need 20+ ns. 
(time cost of System.currentTimeMillis is 50ns)
Actually, event if not hit the cache, 2000ns is not a bottleneck.
And, avoid the overhead of data trans#ission(600bytes) in remotelambda.

## changes in semantics for the user
the serdeLambda didn't change the semantics for the user. the PR keep the Ray.call semantics. Do not conflict with this [#2172](https://github.com/ray-project/ray/issues/2172)

## functions gc
the java func would not be gc, it only gc the MethodId. MethodId is a point to  the called java function.
if miss the MethodId, the MethodId would be reconstruct.

what is the meaning of 'remote lambdas stored' ? I think, there is no need  to store them.

File: java/common/src/main/java/org/ray/util/MethodId.java (hunk 10 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: MethodId is not only for RayFunc, it clould used for all lambda call. the check of RayFunc has add in entrance

Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: MethodId is not only for RayFunc, it clould used for all lambda call. the check of RayFunc has add in entrance

File: java/common/src/main/java/org/ray/util/MethodId.java (hunk 11 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: ok

File: java/common/src/main/java/org/ray/util/MethodId.java (hunk 12 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/MethodId.java]: ok

File: java/common/src/main/java/org/ray/util/LambdaUtils.java (hunk 9 above)
Comment (mylinyuzhi) [reply java/common/src/main/java/org/ray/util/LambdaUtils.java]: MethodId is not only for RayFunc, it clould used for all lambda call. the check of RayFunc has add in entrance

**Review by pcmoritz — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**